src/agent/discovery_proxy.cpp

```diff
@@ -0,0 +1,369 @@
+            {
+                mMdnsPublisher.SubscribeService(serviceName, instanceName);
+            }
+            else
+            {
+                mMdnsPublisher.SubscribeHost(hostName);
+            }
+        }
+    }
+
+exit:
+    if (error != OTBR_ERROR_NONE)
+    {
+        otbrLog(OTBR_LOG_WARNING, "[dnssd discovery proxy] failed to subscribe %s: %s", fullName.c_str(),
+                otbrErrorString(error));
+    }
+}
+
+void DiscoveryProxy::OnDiscoveryProxyUnsubscribe(void *aContext, const char *aFullName)
+{
+    reinterpret_cast<DiscoveryProxy *>(aContext)->OnDiscoveryProxyUnsubscribe(aFullName);
+}
+
+void DiscoveryProxy::OnDiscoveryProxyUnsubscribe(const char *aFullName)
+{
+    std::string                    fullName(aFullName);
+    std::string                    instanceName, serviceName, hostName, domain;
+    otbrError                      error = OTBR_ERROR_NONE;
+    MdnsSubscriptionList::iterator it;
+    DnsNameType                    nameType = GetDnsNameType(fullName);
+
+    otbrLog(OTBR_LOG_INFO, "[dnssd discovery proxy] unsubscribe: %s", fullName.c_str());
+
+    switch (nameType)
+    {
+    case kDnsNameTypeService:
+        SuccessOrExit(error = SplitFullServiceName(fullName, serviceName, domain));
+        break;
+    case kDnsNameTypeInstance:
+        SuccessOrExit(error = SplitFullServiceInstanceName(fullName, instanceName, serviceName, domain));
+        break;
+    case kDnsNameTypeHost:
+        SuccessOrExit(error = SplitFullHostName(fullName, hostName, domain));
+        break;
+    default:
+        ExitNow(error = OTBR_ERROR_NOT_IMPLEMENTED);
+    }
+
+    it = std::find_if(mSubscriptions.begin(), mSubscriptions.end(), [&](const MdnsSubscription &aSubscription) {
+        return aSubscription.Matches(instanceName, serviceName, hostName, domain);
+    });
+
+    VerifyOrExit(it != mSubscriptions.end(), error = OTBR_ERROR_NOT_FOUND);
+
+    {
+        MdnsSubscription &subscription = *it;
+
+        subscription.mSubscriptionCount--;
+        assert(subscription.mSubscriptionCount >= 0);
+
+        if (subscription.mSubscriptionCount == 0)
+        {
+            mSubscriptions.erase(it);
+        }
+
+        otbrLog(OTBR_LOG_DEBUG, "[dnssd discovery proxy] service subscriptions: %sx%d", it->ToString().c_str(),
+                it->mSubscriptionCount);
+
+        if (GetServiceSubscriptionCount(instanceName, serviceName) == 0)
+        {
+            if (subscription.mHostName.empty())
+            {
+                mMdnsPublisher.UnsubscribeService(serviceName, instanceName);
+            }
+            else
+            {
+                mMdnsPublisher.UnsubscribeHost(hostName);
+            }
+        }
+    }
+exit:
+    if (error != OTBR_ERROR_NONE)
+    {
+        otbrLog(OTBR_LOG_WARNING, "[dnssd discovery proxy] failed to unsubscribe %s.%s: %s",
+                instanceName.empty() ? "ANY" : instanceName.c_str(), aFullName, otbrErrorString(error));
```

Comment:
```suggestion
        otbrLog(OTBR_LOG_WARNING, "[dnssd discovery proxy] failed to unsubscribe %s: %s", aFullName, otbrErrorString(error));
```

`aFullName` should already include the instance name, right?